### PR TITLE
cryptsetup: add 'headless' parameter to skip password/pin query, allow pin-less enroll on FIDO2, support user presence/verification flags

### PR DESCRIPTION
**Pull request by bluca**
On headless setups, in case other methods fail, asking for a password
is not useful as there are no users on the terminal, and generates
unwanted noise. Add a parameter to skip it.

### CODE REVIEW

**Comment by poettering** [comment]
maybe call this headless? or non-interactive? i mean, that's what it really is about, that no user interaction shall take place here.

we also ask for the pkcs11 PIN interactively (see `src/cryptsetup/cryptsetup-pkcs11.c`) and FIDO2 PIN (see `src/shared/libfido2-util.c`) which should probably be turned off too by this, no?

**Comment by bluca** [comment]
> maybe call this headless? or non-interactive? i mean, that's what it really is about, that no user interaction shall take place here.
> 
> we also ask for the pkcs11 PIN interactively (see `src/cryptsetup/cryptsetup-pkcs11.c`) and FIDO2 PIN (see `src/shared/libfido2-util.c`) which should probably be turned off too by this, no?

Done - but I don't have the setup to test it.

**Comment by bluca** [comment]
> > maybe call this headless? or non-interactive? i mean, that's what it really is about, that no user interaction shall take place here.
> > we also ask for the pkcs11 PIN interactively (see `src/cryptsetup/cryptsetup-pkcs11.c`) and FIDO2 PIN (see `src/shared/libfido2-util.c`) which should probably be turned off too by this, no?
> 
> Done - but I don't have the setup to test it.

The second commit fixes  FIDO2 + PIN. Tested with 'headless', works as expected.

File: src/cryptenroll/cryptenroll-fido2.c [hunk 1]
```diff
@@ -11,7 +11,8 @@ int enroll_fido2(
                 struct crypt_device *cd,
                 const void *volume_key,
                 size_t volume_key_size,
-                const char *device) {
+                const char *device,
+                bool lock_with_pin) {
```
Comment (poettering) [inline src/cryptenroll/cryptenroll-fido2.c]: if you add this, i'd prefer if we'd hook up all three of "clientPin", "up" and "uv" with user configuration and store the setting in JSON, a the same time (and then make this a flags param)

"up" and "uv" should probably encoded as tristate in JSON, i.e. true/false/null, apping to libfido2's FIDO_OPT_TRUE/FIDO_OPT_FALSE/FIDO_OPT_OMIT

File: src/cryptenroll/cryptenroll-fido2.c [hunk 2]
```diff
@@ -75,7 +77,8 @@ int enroll_fido2(
                                        JSON_BUILD_PAIR("keyslots", JSON_BUILD_ARRAY(JSON_BUILD_STRING(keyslot_as_string))),
                                        JSON_BUILD_PAIR("fido2-credential", JSON_BUILD_BASE64(cid, cid_size)),
                                        JSON_BUILD_PAIR("fido2-salt", JSON_BUILD_BASE64(salt, salt_size)),
-                                       JSON_BUILD_PAIR("fido2-rp", JSON_BUILD_STRING("io.systemd.cryptsetup"))));
+                                       JSON_BUILD_PAIR("fido2-rp", JSON_BUILD_STRING("io.systemd.cryptsetup")),
+                                       JSON_BUILD_PAIR("fido2-pin-required", JSON_BUILD_BOOLEAN(lock_with_pin))));
```
Comment (poettering) [inline src/cryptenroll/cryptenroll-fido2.c]: hmm, fido2 spec calls this "clientPn". Maybe we should stick to that naming, and call this "fido2-clientPin".

File: src/cryptenroll/cryptenroll.c [hunk 3]
```diff
@@ -82,6 +83,7 @@ static int help(void) {
                "\n%sEnroll a security token or authentication credential to a LUKS volume.%s\n\n"
                "  -h --help            Show this help\n"
                "     --version         Show package version\n"
+               "     --lock-with-pin   Require entering a PIN to unlock the volume (FIDO2)\n"
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: should have "fido2" in its name, and i would prefer an explicit option for this:

--fido2-with-client-pin=yes|no

File: src/cryptenroll/cryptenroll-fido2.c [hunk 4]
```diff
@@ -78,7 +79,8 @@ int enroll_fido2(
                                        JSON_BUILD_PAIR("fido2-credential", JSON_BUILD_BASE64(cid, cid_size)),
                                        JSON_BUILD_PAIR("fido2-salt", JSON_BUILD_BASE64(salt, salt_size)),
                                        JSON_BUILD_PAIR("fido2-rp", JSON_BUILD_STRING("io.systemd.cryptsetup")),
-                                       JSON_BUILD_PAIR("fido2-pin-required", JSON_BUILD_BOOLEAN(lock_with_pin))));
+                                       JSON_BUILD_PAIR("fido2-pin-required", JSON_BUILD_BOOLEAN(lock_with_pin)),
+                                       JSON_BUILD_PAIR("fido2-presence-required", JSON_BUILD_BOOLEAN(true))));
```
Comment (poettering) [inline src/cryptenroll/cryptenroll-fido2.c]: i'd also stick to fido2 spec naming, i.e. "fido2-up"

File: src/cryptenroll/cryptenroll-fido2.c (hunk 2 above)
Comment (bluca) [reply src/cryptenroll/cryptenroll-fido2.c]: ok, but I'll leave the suffix `-required`, as it's not storing the actual pin, but just a flag

**Comment by bluca** [comment]
@poettering ok changed as requested - now there are flags for pin, up and uv. Pin and up are enabled by default, uv is disabled by default in cryptenroll as tokens with biometric scans are still not very common. PTAL!

**Comment by bluca** [comment]
(also I left it as-is in homed/homectl, as I don't use it and I am not familiar with it, I hope you can fill in those gaps as a follow-up?)

File: man/crypttab.xml [hunk 5]
```diff
@@ -514,6 +514,13 @@
         user is queried for a password indefinitely.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>headless=</option></term>
+
+        <listitem><para>Takes a boolean argument, defaults to false. If true, never query interactively
+        for the password/PIN. Useful for headless systems.</para></listitem>
```
Comment (keszybz) [inline man/crypttab.xml:521]: Why is this a property of the crypttab entry? It seems to be a property of the system as a whole. If I lift the disk and attach it to my laptop, the device suddenly stops being "headless" and we can query for a pin, but I want to use the same old crypttab entry to configure the device.

File: src/cryptsetup/cryptsetup-fido2.c [hunk 6]
```diff
@@ -88,6 +89,9 @@ int acquire_fido2_key(
 
                 pins = strv_free_erase(pins);
 
+                if (headless)
+                        return log_error_errno(SYNTHETIC_ERRNO(EPERM), "PIN querying disabled via 'headless' option. Use the 'PIN' environment variable.");
```
Comment (keszybz) [inline src/cryptsetup/cryptsetup-fido2.c]: EUNATCH or EOWNERDEAD?

File: src/cryptsetup/cryptsetup-fido2.c [hunk 7]
```diff
@@ -176,6 +183,17 @@ int find_fido2_auto_data(
                         if (!rp)
                                 return log_oom();
                 }
+
+                w = json_variant_by_key(v, "fido2-clientPin-required");
+                if (w) {
+                        /* The "clientPin-required" field is optional. */
```
Comment (keszybz) [inline src/cryptsetup/cryptsetup-fido2.c]: fido2-

**Review by keszybz — COMMENTED** [review]
When reading this, I'm somewhat confused by the lack of distinction of "a headless system where we cannot query the user" with "a device for which we don't want to query the user". The former would be a property of the system, the latter a per-device thing. The series implements the latter, but describes and uses terminology that is fit for the former.

I'm also confused by the overall logic in patch 2. I would expect that whether a pin is required or not is a property configured on the device. I.e. when initially setting up the device, say "require pin = true|false", and then when unlocking, specify the pin or not accordingly. But having the unlocking function *provide* the information that pin is not required feels very strange.

EDIT: I read the comments on the original issue, and I think I see what is happening. But please explain the logic in the commit message.

**Comment by bluca** [comment]
> When reading this, I'm somewhat confused by the lack of distinction of "a headless system where we cannot query the user" with "a device for which we don't want to query the user". The former would be a property of the system, the latter a per-device thing. The series implements the latter, but describes and uses terminology that is fit for the former.

The per-system idea makes sense, but I'm not sure how I would model/implement it in practice?

> I'm also confused by the overall logic in patch 2. I would expect that whether a pin is required or not is a property configured on the device. I.e. when initially setting up the device, say "require pin = true|false", and then when unlocking, specify the pin or not accordingly. But having the unlocking function _provide_ the information that pin is not required feels very strange.
> 
> EDIT: I read the comments on the original issue, and I think I see what is happening. But please explain the logic in the commit message.

Yeah the device can _mandate_ using a pin, or can merely provide the user with a choice for every token. So we leave it open with a command line option on enrollment in case it's optional, but error out immediately if it's mandatory. I'll update the commit message.

**Comment by keszybz** [comment]
> > When reading this, I'm somewhat confused by the lack of distinction of "a headless system where we cannot query the user" with "a device for which we don't want to query the user". The former would be a property of the system, the latter a per-device thing. The series implements the latter, but describes and uses terminology that is fit for the former.
> 
> The per-system idea makes sense, but I'm not sure how I would model/implement it in practice?

*I* am not sure if it makes sense. But if it does, then `machine-info` would seem the appropriate place.

**Comment by bluca** [comment]
> > > When reading this, I'm somewhat confused by the lack of distinction of "a headless system where we cannot query the user" with "a device for which we don't want to query the user". The former would be a property of the system, the latter a per-device thing. The series implements the latter, but describes and uses terminology that is fit for the former.
> > 
> > 
> > The per-system idea makes sense, but I'm not sure how I would model/implement it in practice?
> 
> _I_ am not sure if it makes sense. But if it does, then `machine-info` would seem the appropriate place.

Either works for me, I don't mind too much - so a new `HEADLESS=yes|no` variable? Or matching on `embedded` or `server` `CHASSIS`? The former seems a bit over-specific, while the latter could be an unexpected change for existing users

**Comment by keszybz** [comment]
@poettering opinions?

File: src/cryptsetup/cryptsetup.c [hunk 8]
```diff
@@ -381,6 +382,8 @@ static int parse_one_option(const char *option) {
 
         } else if (streq(option, "try-empty-password"))
                 arg_try_empty_password = true;
+        else if (streq(option, "headless"))
+                arg_headless = true;
```
Comment (poettering) [inline src/cryptsetup/cryptsetup.c:395]: you document this in a way this expects a bool argument, but here you implement it without argument. I think it should support both, but document with a bool argument. i.e. like try-empty-password does it. in other words. please also parse headless= here with an argument

File: src/cryptsetup/cryptsetup.c [hunk 9]
```diff
@@ -532,6 +535,9 @@ static int get_password(
         assert(src);
         assert(ret);
 
+        if (arg_headless)
+                return log_error_errno(SYNTHETIC_ERRNO(EPERM), "Password querying disabled via 'headless' option.");
```
Comment (poettering) [inline src/cryptsetup/cryptsetup.c]: we probably should return a common error for this. In the fido2 case you return EUNATCH, but you could also use ENOMEDIUM? ENOPKG? something that isn't use otherwise by the cryptsetup/luks/kernel/pkc11/fido2 stuff here so far...

File: src/shared/pkcs11-util.c [hunk 10]
```diff
@@ -247,7 +248,9 @@ int pkcs11_token_login(
                         string_erase(e);
                         if (unsetenv("PIN") < 0)
                                 return log_error_errno(errno, "Failed to unset $PIN: %m");
-                } else {
+                } else if (headless)
+                        return log_error_errno(SYNTHETIC_ERRNO(EPERM), "PIN querying disabled via 'headless' option. Use the 'PIN' environment variable.");
```
Comment (poettering) [inline src/shared/pkcs11-util.c]: as above. let's use a common error

File: src/cryptsetup/cryptsetup-fido2.c [hunk 11]
```diff
@@ -88,6 +89,9 @@ int acquire_fido2_key(
 
                 pins = strv_free_erase(pins);
 
+                if (headless)
+                        return log_error_errno(SYNTHETIC_ERRNO(EUNATCH), "PIN querying disabled via 'headless' option. Use the 'PIN' environment variable.");
```
Comment (poettering) [inline src/cryptsetup/cryptsetup-fido2.c]: i'd denote env var names always with a preceding $, to clarfiy they are env vars.

File: src/cryptenroll/cryptenroll.c [hunk 12]
```diff
@@ -88,6 +89,8 @@ static int help(void) {
                "                       Specify PKCS#11 security token URI\n"
                "     --fido2-device=PATH\n"
                "                       Enroll a FIDO2-HMAC security token\n"
+               "     --fido2-with-client-pin=[yes|no]\n"
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: we usually write `=BOOL` instead of `=[yes|no]` in help texts

File: src/cryptsetup/cryptsetup.c [hunk 13]
```diff
@@ -750,14 +751,19 @@ static int attach_luks_or_plain_or_bitlk_by_fido2(
...
+                                &pin_required);
 
                 if (IN_SET(r, -ENOTUNIQ, -ENXIO))
                         return log_debug_errno(SYNTHETIC_ERRNO(EAGAIN),
                                                "Automatic FIDO2 metadata discovery was not possible because missing or not unique, falling back to traditional unlocking.");
                 if (r < 0)
                         return r;
 
+                if (pin_required && arg_headless)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
```
Comment (poettering) [inline src/cryptsetup/cryptsetup.c]: we probably should use that same error code here that is used for "pin/password needed, but headless is on", see other comments.

File: man/systemd-cryptenroll.xml [hunk 14]
```diff
@@ -132,6 +132,14 @@
         enter a PIN when unlocking the volume. Defaults to <literal>yes</literal>.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--fido2-with-user-presence=</option><replaceable>[yes|no]</replaceable></term>
+
+        <listitem><para>When enrolling a FIDO2 security token, controls whether to require the user to
+        verify presence (tap the token) when unlocking the volume. Defaults to <literal>yes</literal>.
```
Comment (poettering) [inline man/systemd-cryptenroll.xml]: maybe mention that fido2 spec calls this "up". i.e. add `(the FIDO2 <literal>up</literal> feature)`,

File: src/cryptsetup/cryptsetup-fido2.c [hunk 15]
```diff
@@ -81,7 +82,7 @@ int acquire_fido2_key(
                                         salt, salt_size,
                                         cid, cid_size,
                                         pins,
-                                        /* up= */ true,
+                                        presence_required,
```
Comment (poettering) [inline src/cryptsetup/cryptsetup-fido2.c]: maybe say "user_presence_requires" and add `/* FIDO2 "up" */` in comments

File: src/cryptsetup/cryptsetup-fido2.c [hunk 16]
```diff
@@ -194,6 +198,17 @@ int find_fido2_auto_data(
...
                 }
+
+                w = json_variant_by_key(v, "fido2-up-required");
+                if (w) {
+                        /* The "fido2-up-required" field is optional. */
+
+                        if (!json_variant_is_boolean(w))
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "FIDO2 token data's 'fido2-up-required' field is not a boolean.");
+
```
Comment (poettering) [inline src/cryptsetup/cryptsetup-fido2.c:205]: maybe ad "else" here

File: src/shared/libfido2-util.c [hunk 17]
```diff
@@ -256,6 +256,11 @@ static int fido2_use_hmac_hash_specific_token(
                                        "PIN required to unlock, but FIDO2 device %s does not support it.",
                                        path);
 
+        if (!has_up && up_required)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
```
Comment (poettering) [inline src/shared/libfido2-util.c]: maybe for both these cases use EHWPOISON, to indicate the hw doesn't meet our expectations

File: src/shared/libfido2-util.c [hunk 18]
```diff
@@ -521,6 +531,11 @@ int fido2_generate_hmac_hash(
                                        "Requested to lock with PIN, but FIDO2 device %s does not support it.",
                                        device);
 
+        if (!has_up && lock_with_presence)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Requested to lock with user presence, but FIDO2 device %s does not support it.",
+                                       device);
+
```
Comment (poettering) [inline src/shared/libfido2-util.c:541]: we should set sym_fido_assert_set_up() based on lock_with_presence I guess, further down, so that the assert config matches exactly what is encoded in the JSON data

File: src/cryptenroll/cryptenroll-fido2.c [hunk 19]
```diff
@@ -13,7 +13,8 @@ int enroll_fido2(
                 size_t volume_key_size,
                 const char *device,
                 bool lock_with_pin,
-                bool lock_with_up) {
+                bool lock_with_up,
+                bool lock_with_uv) {
```
Comment (poettering) [inline src/cryptenroll/cryptenroll-fido2.c]: I'd prefer if this three bools would be replaced via some new flags argument, so that this becomes `Fido2EnrollFlags lock_with` or so

File: src/cryptsetup/cryptsetup-fido2.c [hunk 20]
```diff
@@ -209,6 +214,17 @@ int find_fido2_auto_data(
...
                 }
+
+                w = json_variant_by_key(v, "fido2-uv-required");
+                if (w) {
+                        /* The "fido2-uv-required" field is optional. */
+
+                        if (!json_variant_is_boolean(w))
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "FIDO2 token data's 'fido2-uv-required' field is not a boolean.");
+
```
Comment (poettering) [inline src/cryptsetup/cryptsetup-fido2.c:216]: else

File: src/shared/libfido2-util.c [hunk 21]
```diff
@@ -220,13 +222,14 @@ static int fido2_use_hmac_hash_specific_token(
                 char **pins,
                 bool up_required, /* user presence required */
                 bool pin_required, /* client pin required */
+                bool uv_required, /* user verification required */
```
Comment (poettering) [inline src/shared/libfido2-util.c]: these three should then reuse the flags type, `Fido2EnrollFlags required`

File: src/shared/libfido2-util.c [hunk 22]
```diff
@@ -323,6 +331,21 @@ static int fido2_use_hmac_hash_specific_token(
                 else
                         r = sym_fido_dev_get_assert(d, a, NULL); /* try without pin but with up now */
         }
+        if (uv_required && has_uv) {
+                r = sym_fido_assert_set_uv(a, FIDO_OPT_TRUE);
```
Comment (poettering) [inline src/shared/libfido2-util.c]: let's just invoke this beforehand with the value encoded in the JSON data. i.e. if we know exactly what is configured we don#t have to try multiple times here

**Review by poettering — CHANGES_REQUESTED** [review]
looks excellent, some points though

File: src/cryptsetup/cryptsetup-fido2.c (hunk 16 above)
Comment (bluca) [reply src/cryptsetup/cryptsetup-fido2.c:205]: sorry, not following - there's a return, and we usually don't use `else` to reduce indentation, no?

File: src/shared/libfido2-util.c (hunk 18 above)
Comment (bluca) [reply src/shared/libfido2-util.c:541]: it is already doing that:

```
        if (r == FIDO_ERR_UP_REQUIRED || (lock_with_presence && has_up)) {

                if (!has_up)
                        log_warning("Weird, device asked for User Presence check, but does not advertise it as feature. Ignoring.");

                r = sym_fido_assert_set_up(a, FIDO_OPT_TRUE);
```

Or did you mean something else?

Comment (poettering) [reply src/shared/libfido2-util.c:541]: What I am saying is that we should tweak up/uv/clientPin anymore after we tried once. we should store precisely what up/uv/clientPin was set to when we enrolled it and then rebuild the assertion from that, and try exactly once with that data, and not try any other config anymore if that fails

File: man/systemd-cryptenroll.xml [hunk 23]
```diff
@@ -125,6 +125,13 @@
         <filename>/etc/crypttab</filename> line.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--fido2-with-client-pin=</option><replaceable>[yes|no]</replaceable></term>
```
Comment (poettering) [inline man/systemd-cryptenroll.xml]: [yes|no] → BOOL

File: src/cryptenroll/cryptenroll.c [hunk 24]
```diff
@@ -144,6 +149,18 @@ static int parse_argv(int argc, char *argv[]) {
...
                         return version();
 
+                case ARG_FIDO2_WITH_PIN: {
+                        bool lock_with_pin;
+
+                        r = parse_boolean_argument("--fido2-with-client-pin=", optarg, &lock_with_pin);
+                        if (r < 0)
+                                return r;
+
+                        arg_fido2_lock_with = UPDATE_FLAG(arg_fido2_lock_with, FIDO2ENROLL_PIN, lock_with_pin);
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]:
```c
SET_FLAG(arg_fido2_lock_with, FIDO2ENROLL_PIN, lock_with_pin); 
```

works too and is a tiny bit shorter

File: src/cryptsetup/cryptsetup-fido2.c [hunk 25]
```diff
@@ -176,6 +183,17 @@ int find_fido2_auto_data(
...
+                w = json_variant_by_key(v, "fido2-clientPin-required");
+                if (w) {
+                        /* The "fido2-clientPin-required" field is optional. */
+
+                        if (!json_variant_is_boolean(w))
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "FIDO2 token data's 'fido2-clientPin-required' field is not a boolean.");
+
+                        required = UPDATE_FLAG(required, FIDO2ENROLL_PIN, json_variant_boolean(w));
```
Comment (poettering) [inline src/cryptsetup/cryptsetup-fido2.c]: SET_FLAG()

File: src/shared/libfido2-util.c [hunk 26]
```diff
@@ -301,7 +307,12 @@ static int fido2_use_hmac_hash_specific_token(
 
                 log_info("Security token requires user presence.");
 
-                r = sym_fido_dev_get_assert(d, a, NULL); /* try without pin but with up now */
+                /* If the client pin is required, do not attempt to skip it, or it will work but give
+                 * a different secret */
+                if (FLAGS_SET(required, FIDO2ENROLL_PIN))
```
Comment (poettering) [inline src/shared/libfido2-util.c]: given that we *know* that we need a PIN we should never try without one. previously we didn't store the fact that we need a PIN in the JSON data. but we do now, hence we really shouldn't bother at all with settings that don't matc what we know from the json data.

File: man/systemd-cryptenroll.xml [hunk 27]
```diff
@@ -132,6 +132,15 @@
         enter a PIN when unlocking the volume. Defaults to <literal>yes</literal>.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--fido2-with-user-presence=</option><replaceable>[yes|no]</replaceable></term>
```
Comment (poettering) [inline man/systemd-cryptenroll.xml]: [yes|no] → BOOL

File: src/cryptenroll/cryptenroll.c [hunk 28]
```diff
@@ -161,6 +165,18 @@ static int parse_argv(int argc, char *argv[]) {
...
                 }
 
+                case ARG_FIDO2_WITH_UP: {
+                        bool lock_with_up;
+
+                        r = parse_boolean_argument("--fido2-with-user-presence=", optarg, &lock_with_up);
+                        if (r < 0)
+                                return r;
+
+                        arg_fido2_lock_with = UPDATE_FLAG(arg_fido2_lock_with, FIDO2ENROLL_UP, lock_with_up);
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: SET_FLAG

File: src/cryptsetup/cryptsetup-fido2.c (hunk 16 above)
Comment (poettering) [reply src/cryptsetup/cryptsetup-fido2.c:205]: ignore me. i am confused.

File: src/cryptsetup/cryptsetup-fido2.c [hunk 29]
```diff
@@ -194,6 +194,17 @@ int find_fido2_auto_data(
...
+
+                w = json_variant_by_key(v, "fido2-up-required");
+                if (w) {
+                        /* The "fido2-up-required" field is optional. */
+
+                        if (!json_variant_is_boolean(w))
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "FIDO2 token data's 'fido2-up-required' field is not a boolean.");
+
+                        required = UPDATE_FLAG(required, FIDO2ENROLL_UP, json_variant_boolean(w));
```
Comment (poettering) [inline src/cryptsetup/cryptsetup-fido2.c]: SET_FLAG

File: src/shared/libfido2-util.c [hunk 30]
```diff
@@ -627,6 +659,11 @@ int fido2_generate_hmac_hash(
         if (r == FIDO_ERR_PIN_AUTH_BLOCKED)
                 return log_notice_errno(SYNTHETIC_ERRNO(EPERM),
                                         "Token PIN is currently blocked, please remove and reinsert token.");
+#ifdef FIDO_ERR_UV_BLOCKED
+        if (r == FIDO_ERR_UV_BLOCKED)
+                return log_notice_errno(SYNTHETIC_ERRNO(EPERM),
+                                        "Token verification is currently blocked, please remove and reinsert token.");
+#endif
```
Comment (poettering) [inline src/shared/libfido2-util.c:645]: maybe split this into separate commit

**Review by poettering — CHANGES_REQUESTED** [review]
This looks pretty good, but if we have the JSON data and shall assert, then let's configure the assertion 1:1 based on the JSON data, and make that the only thing we try. Previously we'd try a couple of different ways to assert, because we didn't store up/uv/clientPin, and we wanted to figure it out. But that's unnecessary now that we store everything anyway.

File: src/shared/libfido2-util.c (hunk 18 above)
Comment (bluca) [reply src/shared/libfido2-util.c:541]: ok, got it now, thanks

**Comment by bluca** [comment]
> This looks pretty good, but if we have the JSON data and shall assert, then let's configure the assertion 1:1 based on the JSON data, and make that the only thing we try. Previously we'd try a couple of different ways to assert, because we didn't store up/uv/clientPin, and we wanted to figure it out. But that's unnecessary now that we store everything anyway.

Done, PTAL!

Also please check this comment from @keszybz - I do not mind one way or another, both work for me:

https://github.com/systemd/systemd/pull/19298#issuecomment-821053720

**Comment by bluca** [comment]
bionic-i386 is the usual timeout in `TEST-24-CRYPTSETUP` and centos-arch is `TEST-58-REPART`, unrelated

**Comment by 239** [comment]
Will this be included in the next release or is there still something missing?

File: src/shared/libfido2-util.c [hunk 31]
```diff
@@ -516,6 +510,11 @@ int fido2_generate_hmac_hash(
                                        "Requested to lock with PIN, but FIDO2 device %s does not support it.",
                                        device);
 
+        if (!has_up && FLAGS_SET(lock_with, FIDO2ENROLL_UP))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Requested to lock with user presence, but FIDO2 device %s does not support it.",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: "Locking with user presence test requested, but …"

File: src/shared/libfido2-util.c [hunk 32]
```diff
@@ -652,32 +651,27 @@ int fido2_generate_hmac_hash(
...
 
+        log_info("Generating secret key on FIDO2 security token.");
+
         if (has_up) {
-                r = sym_fido_assert_set_up(a, FIDO_OPT_FALSE);
+                r = sym_fido_assert_set_up(a, FLAGS_SET(lock_with, FIDO2ENROLL_UP) ? FIDO_OPT_TRUE : FIDO_OPT_FALSE);
                 if (r != FIDO_OK)
                         return log_error_errno(SYNTHETIC_ERRNO(EIO),
-                                               "Failed to turn off FIDO2 assertion user presence: %s", sym_fido_strerr(r));
+                                               "Failed to %s FIDO2 assertion user presence: %s",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: "Failed to %s FIDO2 user presence test"

File: src/shared/libfido2-util.c [hunk 33]
```diff
@@ -652,32 +651,27 @@ int fido2_generate_hmac_hash(
...
+                                               enable_disable(FLAGS_SET(lock_with, FIDO2ENROLL_UP)),
+                                               sym_fido_strerr(r));
+
+                if (FLAGS_SET(lock_with, FIDO2ENROLL_UP))
+                        log_notice("%s%sIn order to allow secret key generation, please verify presence on security token.",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: verify → confirm ? I think it's more natural to say "confirm who you are" rather than "verify who you are" when it's just a simple check. Especially that the next patch adds "verification" in a completely different sense.

File: src/shared/libfido2-util.c [hunk 34]
```diff
@@ -652,32 +651,27 @@ int fido2_generate_hmac_hash(
...
-
-                log_notice("%s%sIn order to allow secret key generation, please verify presence on security token.",
-                           emoji_enabled() ? special_glyph(SPECIAL_GLYPH_TOUCH) : "",
-                           emoji_enabled() ? " " : "");
-
-                r = sym_fido_dev_get_assert(d, a, FLAGS_SET(lock_with, FIDO2ENROLL_PIN) ? used_pin : NULL);
-        }
+        if (r == FIDO_ERR_UP_REQUIRED && !FLAGS_SET(lock_with, FIDO2ENROLL_UP))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Requested to lock without user presence, but FIDO2 device %s requires it.",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: "Locking without user presence test requested, but …"

File: src/shared/libfido2-util.c [hunk 35]
```diff
@@ -256,6 +255,11 @@ static int fido2_use_hmac_hash_specific_token(
                                        "PIN required to unlock, but FIDO2 device %s does not support it.",
                                        path);
 
+        if (!has_up && FLAGS_SET(required, FIDO2ENROLL_UP))
+                return log_error_errno(SYNTHETIC_ERRNO(EHWPOISON),
+                                       "User presence required to unlock, but FIDO2 device %s does not support it.",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: "User presence *test* required to unlock"

File: src/shared/libfido2-util.c [hunk 36]
```diff
@@ -285,30 +289,20 @@ static int fido2_use_hmac_hash_specific_token(
...
-
-                r = sym_fido_assert_set_up(a, FIDO_OPT_TRUE);
+        if (has_up) {
+                r = sym_fido_assert_set_up(a, FLAGS_SET(required, FIDO2ENROLL_UP) ? FIDO_OPT_TRUE : FIDO_OPT_FALSE);
                 if (r != FIDO_OK)
                         return log_error_errno(SYNTHETIC_ERRNO(EIO),
-                                       "Failed to set FIDO2 assertion user presence: %s", sym_fido_strerr(r));
-
-                log_info("Security token requires user presence.");
+                                               "Failed to %s FIDO2 assertion user presence: %s",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: "Failed to %s FIDO2 user presence test: %s"

File: src/shared/libfido2-util.c [hunk 37]
```diff
@@ -303,6 +310,18 @@ static int fido2_use_hmac_hash_specific_token(
                         log_info("User presence required to unlock.");
         }
 
+        if (has_uv) {
+                r = sym_fido_assert_set_uv(a, FLAGS_SET(required, FIDO2ENROLL_UV) ? FIDO_OPT_TRUE : FIDO_OPT_FALSE);
+                if (r != FIDO_OK)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                               "Failed to %s FIDO2 assertion user verification: %s",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: "Failed to %s FIDO2 user verification: %s"

File: src/shared/libfido2-util.c [hunk 38]
```diff
@@ -515,6 +534,11 @@ int fido2_generate_hmac_hash(
                                        "Requested to lock with user presence, but FIDO2 device %s does not support it.",
                                        device);
 
+        if (!has_uv && FLAGS_SET(lock_with, FIDO2ENROLL_UV))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Requested to lock with user verification, but FIDO2 device %s does not support it.",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: "Locking with user verification requested, but …"

File: src/shared/libfido2-util.c [hunk 39]
```diff
@@ -667,6 +691,20 @@ int fido2_generate_hmac_hash(
                                    emoji_enabled() ? " " : "");
         }
 
+        if (has_uv) {
+                r = sym_fido_assert_set_uv(a, FLAGS_SET(lock_with, FIDO2ENROLL_UV) ? FIDO_OPT_TRUE : FIDO_OPT_FALSE);
+                if (r != FIDO_OK)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                               "Failed to %s FIDO2 assertion user verification: %s",
```
Comment (keszybz) [inline src/shared/libfido2-util.c]: "Failed to %s FIDO user verification: %s"

File: src/shared/libfido2-util.c [hunk 40]
```diff
@@ -633,6 +638,11 @@ int fido2_generate_hmac_hash(
         if (r == FIDO_ERR_PIN_AUTH_BLOCKED)
                 return log_notice_errno(SYNTHETIC_ERRNO(EPERM),
                                         "Token PIN is currently blocked, please remove and reinsert token.");
+#ifdef FIDO_ERR_UV_BLOCKED
+        if (r == FIDO_ERR_UV_BLOCKED)
+                return log_notice_errno(SYNTHETIC_ERRNO(EPERM),
+                                        "Token verification is currently blocked, please remove and reinsert token.");
```
Comment (keszybz) [inline src/shared/libfido2-util.c:644]: It bugs me a bit that those messages are different here and above…

Comment (bluca) [reply src/shared/libfido2-util.c:644]: sorry, compared to which ones?

Comment (keszybz) [reply src/shared/libfido2-util.c:644]: Above in this patch. Try this view: https://github.com/systemd/systemd/pull/19298/files/e17429a2998e3600d9f0874bb65d571018fe2335..c5fd3f8da40409ea9bca2c068ce8100f26c8ef2f#diff-bf30439d1f70cf4825bc72a9707c7c157000b8dcc6c2cd4d936aea7519536f4a

Comment (bluca) [reply src/shared/libfido2-util.c:644]: Yeah, the original ones are different too, it seemed better to keep the messages consistent with the area

**Comment by bluca** [comment]
Updated, thanks for the review

**Comment by keszybz** [comment]
Let's merge this when CI passes. I don't think yet-another review will help.

**Comment by bluca** [comment]
rebased to pick up unrelated CI fixes, no other changes

**Comment by bluca** [comment]
CentOS/Arch is failing at `TEST-51-ISSUE-16115` which feels unrelated - @mrc0mmand have you seen that before?

**Comment by mrc0mmand** [comment]
> CentOS/Arch is failing at `TEST-51-ISSUE-16115` which feels unrelated - @mrc0mmand have you seen that before?

I think I've seen that one only once, but it's definitely not related:

```
May 07 23:28:22 H testsuite-51.sh[36]: ++ systemctl show testsuite-51-repro-1 -P NRestarts
May 07 23:28:47 H testsuite-51.sh[36]: Failed to get properties: Connection timed out
May 07 23:28:47 H testsuite-51.sh[27]: + [[ '' == \0 ]]
```

**Comment by bluca** [comment]
Thanks! Merging then
